### ISSUE TITLE
Big slack dump

### ISSUE THREAD

**Issue by oxinabox** [JuliaTakingFittingAPIsSeriously/Roadmap#3]
This is the big dump of stuff from slack.
Close this once it is in a sane format in another issues

-----------------------

Lyndon White [2:27 PM]
Idea.
We create a Base package.
Call it ModelingBase.jl or StatsBaseBase.jl
It exports roughly the same things as 
https://github.com/cstjean/ScikitLearnBase.jl/blob/master/docs/API.md
And basically the core of https://github.com/JuliaStats/StatsBase.jl/blob/master/src/statmodels.jl

being `fit`, `fit!`, `predict`, `predict!`, `predict_proba` (which StatsBase doesn't have?)
But unlike ScikitLearnBase.jl it is Column Major.

StatsBase deprends on it,  and deprecates all of its bits to it. (making packages need to update)

Such a ModelingBase.jl would be a light-weight package (unlike StatsBase.jl).
GitHub
cstjean/ScikitLearnBase.jl
ScikitLearnBase.jl - Abstract interface of ScikitLearn.jl
GitHub
JuliaStats/StatsBase.jl
StatsBase.jl - Basic statistics for Julia
If such a package existsed, I would go and make the dozen PRs to all the small ML packages (starting with the 5 from my blog post http://white.ucc.asn.au/2017/12/18/7-Binary-Classifier-Libraries-in-Julia.html; since I've written that code already),
and add to them the functionality, so we can have things not be so variable in their interface
Maybe the function list should include all of `nobs`, `fit`, `fit!`, `predict`, `params`, `params!`
because that would let LearnBase.jl drop from having a StatsBase.jl dependency to having a `ModellingBase.jl` dependency.
Though `nobs` seems a bit off
like `nobs` also seems off for StatsBase
Huh StatsBase uses it to test how many observations were used to train a model.
That is absurd isn't it?

Chris Rackauckas [2:49 PM]
it's not absurd
and these things need to get put together nicely because right now it's a mess.

Eric Forgy [2:50 PM]
Ok. StatsBaseBase.jl is a little absurd :sweat_smile:
(just kidding :blush: )

Lyndon White [2:51 PM]
leaving `nobs` aside, 
does my idea of extracting just a ModelingBase with just the functions that involve fitting some kinda model and applying it to get some kinda prediction out, make sense? (edited)
Would it improve the state of " right now it's a mess."? (edited)

Chris Rackauckas [2:56 PM]
Yes.
Someone has to do it

Chris Rackauckas [2:58 PM]
I wished that it would be Stats.jl but it looks like that definition of stats excludes hypothesis testing when it's done on biological data so it's pretty clear where this is going.

https://github.com/JuliaStats/Stats.jl/pull/4#issuecomment-389562364
GitHub
Add essential packages for statistics by nalimilan · Pull Request #4 · JuliaStats/Stats.jl
This makes the package useful again.

Lyndon White [2:58 PM]
Taking Fitting Seriously
Other notions would be to keep it a very small package, with very few assumptions, and let StatsBase and LearnBase add their assumptions on to it (e.g both of them kinda act like subtyping one of their abstract model types is a thing someone might do. But duck typing is enough)

Chris Rackauckas [3:10 PM]
I think that the general pipeline is what matters
it should be easy to:

1. Get data into a common format (JuliaDB? Dataframes?)
2. Dimensionally reduce the data with the basic methods that everyone uses (PCA, diffusion maps, TSne)
3. Fit whatever model on this data and make predictions
4. Plot the results
5. Cluster on the reduced data (edited)
If you open up a NSC paper (Nature, Science, Cell), or any other high impact journal (any PLOS, ELife, etc.), this is what you'll see
it's straightforward to do in Python and R
and it's pretty clearly a jagged road right now in Julia
I think the first concrete step is to get all of the fitting predictions to be aligned
they should just all be column-major, which is surprisingly not the case right now
Then the same should be done with dimensional reduction

Eric Forgy [3:14 PM]
Yes. imagine what a boost it would be for Julia in general if you guys came tearing up the place with good tools and record breaking benchmarks

Chris Rackauckas [3:14 PM]
and then the dataframe packages should be able to interface directly with how these tools want the data

Chris Rackauckas [3:14 PM]
then Clustering.jl's return type mess needs to get cleaned up
and I think we're good

Eric Forgy [3:15 PM]
I'm cheering from the sidelines. Wish I could do more than that, but constrained at the moment :sweat_smile: (edited)

Lyndon White [3:17 PM]
IYeah so my plan is step 3 on that list.
And to an extent 2 and 5 maybe since they can also be phrases as `fit` and `predict`

Chris Rackauckas [3:18 PM]
need density-based or graphical-based clustering for post-TSne.

Eric Forgy [3:19 PM]
Should turn this into an issue and track it :blush: Good stuff :+1:
Chris Rackauckas
it should be easy to:

1. Get data into a common format (JuliaDB? Dataframes?)
2. Dimensionally reduce the data with the basic methods that everyone uses (PCA, diffusion maps, TSne)
3. Fit whatever model on this data and make predictions
4. Plot the results
5. Cluster on the reduced data
Posted in #machine-learningYesterday at 3:11 PM

Chris Rackauckas [3:19 PM]
TSne.jl is one of the row-major packages I think

Lyndon White [3:25 PM]
It is indeed
It is also licensed such that no one should fork it. (edited)

Lyndon White [3:27 PM]
Its not that it is an evil license it is just that is is crayon (edited)

Chris Rackauckas [3:27 PM]
I haven't checked the license yet since I haven't felt the need to look into ti yet

Lyndon White [5:49 PM]
@evizero @joshday @ararslan   um other people who do stuff with StatsBase / LearnBase /etc
Thoughts on what I said above about separating out `fit` and `predict` and a few other things like that from `StatsBase` into a more compact Baser package?

Christof Stocker [5:51 PM]
right now I am rather neutral since it doesn't really touch my current work and I have other things on my mind

Lyndon White [5:51 PM]
Fair enough

Christof Stocker [5:55 PM]
There have been lots of discussions about this topic all over the place in the past. I think the most fruitful way to make progress is to have some practical downstream use for this "move". So a driving force other than "making the design nicer"
(i am not suggesting that is your motivation, just that this has been one of the main themes in the past (myself included) and obviously that didn't solve it)

Christof Stocker [5:58 PM]
didn't StatsBase dependencies slim down quite a bit?

Christof Stocker [5:59 PM]
thats the "base" package JuliaML is using at its root currently, with the main gripe that it had a binary dependency in the past
which I think to recall is gone

Lyndon White [6:01 PM]
Use case:
A paper I am working with some-one on, calls for a comparason between RandomForest, SVM, and LogisticRegression.
They are working in python, so I said "Just use sklearn and loop over the classifier types"
In julia you can't do that because each classifier type has its own API.
To giving them a common API they should have a common namepace from a FooBase package.
The logical package would be `StatsBase.jl`, because it exports `fit!` and `predict`.
However, StatsBase may very weill be the heaviest weight Base package in the ecosystem. 
So If I go and make a PR on each of the relevant packages, I would expect them to be unhappy at that dependency add
Further I feel like `StatsBase` has all these extra expectations of a model, that not all models meet. THough I guess that isn't so much of an issue. since most will method error immediately

Christof Stocker [6:06 PM]
I think the issue is deeper than a common base package though. Coming up and implementing a common API that all packages should use is quite a challenge. I tend to think nowadays that the way to actually get this done has two phases. first demonstrate in some sort of "caret"-like meta-package that such an API is possible and worth it, and then convince the authors that refactoring the actual packages is worth it.

Lyndon White [6:07 PM]
My plan was to just go and make all the PRs myself
I've written like all the binary ones for that blog post.
So it is just generalising them to multiclass
I could script up a demo that would show looping over all them  (after calling `Pkg.checkout(mybranch)` them and training some task though

Josh Day [7:01 PM]
@oxinabox The idea is that all this package contains is empty function definitions `fit!`, `nobs`, etc.?  I am 100% on board.

Lyndon White [7:01 PM]
yep, and some docs saying Implement it like this,  as Column Major **not** as Row Major (edited)

Josh Day [7:03 PM]
I would argue in favor of `predict` returning probabilities and `classify` returning class, rather than `predict` and `predict_proba`

Lyndon White [7:03 PM]
I am in agreement with that

Josh Day [7:04 PM]
Forcing column/row major might be a hard sell.

Lyndon White [7:05 PM]
It cant't be forced, but it can be documented
and issues can be risen against non-compliant implementations
Maybe an automated testing package? But that perhaps getting ahead of self (edited)
SklearnBase.jl has mostly the right idea, except they are solidly Row Major.
(and I am not so sold on the way the first argument is done as it seems over complicated, as a type that defined the hyperparameters) (edited)

Josh Day [7:16 PM]
Just as a reference, you’ve probably already seen this: https://github.com/JuliaStats/StatsBase.jl/issues/116
GitHub
Fitting in with the Julia ecosystem · Issue #116 · JuliaStats/StatsBase.jl
I&#39;m working on a BayesianLinearRegression.jl module, and I&#39;d like for it to fit in with the Julia ecosystem in a natural way. @johnmyleswhite has already been very helpful regarding general...

Lyndon White [7:20 PM]
I have not
oooh it is a long one
My proposed solution to things that are implement Row major is to expose them as Column Major via transpose (which is now nonallocating anyway, iirc), when overloading the ModellingBase.jl functions.
And not solve the performance issues of existing packages, just the API issues (edited)

Dave F Kleinschmidt [9:02 PM]
replied to a thread:
just catching up here but the formula bits from StatsModels are meant to provide that interface (transform tables to column major matrices for consumption by models)

Chris Rackauckas [9:31 PM]
replied to a thread:
@piever does JuliaDB work with `@formula`s?
@shashi as well
if `fit` et al. can work directly from forumlas and pull from JuliaDB tables at will (and utilize the out-of-core in a good way for batching?) then we'd be in a good place.
@oxinabox row major for what reason? Because of how tables are structured?
so columns as attributes and rows as observations?
It's hard for me to recognize which is "row major" since both directions have meaning

Lyndon White [9:34 PM]
Observations should be the last dimension
Just like distributions.jl, multivariate data.jl etc

Chris Rackauckas [9:35 PM]
so the opposite of table structures?

Lyndon White [9:36 PM]
Maybe?
I guess so
It is the order that leads to fast computation

Chris Rackauckas [9:37 PM]
I think it's fine if there's an known way to use tables
and for example they do a single transpose at the start of their algorithm if they get data from a table
from `@formula`
but integration with tables/databases is essential IMO

Lyndon White [9:39 PM]
Sure. And that integration is part of the bigger picture. It become s easier to build if first everything us using the same function names the same ways.
Beyond even row major and column major is the differences like does predict give probabilities or classes

Dave F Kleinschmidt [9:43 PM]
oh I was confused, i thought column major was what formula did
to keep things together:
Dave F Kleinschmidt
it actually may be more central than it appears, since in order to `predict` from a `fit`ted model based on a formula using held out data you need to keep the formula around to do the transformation again. the current design in StatsModels achieves this by using a wrapper type and delegating the API methods (fit, predict, etc.) to the wrapped model.  This has the advantage that you get formula support "for free" (by defining `fit(m::M, df::DataFrame, f::Formula) where M<:StatisticalModel` or something like that to create the wrapped type), but the downside is that there's no automatic delegation, so any special methods that work on the model type `M` won't work on the wrapped type `DataFrameStatist…` Show more
From a thread in #machine-learningYesterday at 9:40 PM
there are tricky composability things involved with using formulae or any kind of data transformation
as part of model fitting/prediction
AFAIK the way that StatsModels handles this is very different in philosophy from how JuliaDB does.  AFAIK JuliaDB.ML puts the onus on the user to do the data transformations themselves before passing to modeling functions, whereas StatsModels aims for something a little more automagic by overloading `fit`
which is useful because it keeps metadata about the data used to fit the model (e.g., names of predictor features), and it keeps the data-transformer functionality with the model so that `predict` is handled automatically
(and safely)
(e.g., checks whether the same categorical levels are present)
anyway this seems like it deserves discussion in a less ephemeral medium

Chris Rackauckas [9:51 PM]
I agree

Dave F Kleinschmidt [9:51 PM]
I think making a base package that just has `fit`, `predict`, etc. is a good start though

Chris Rackauckas [9:51 PM]
I think that JuliaDB devs should weigh in on how they plan to do this

Lyndon White [9:51 PM]
Sounds useful and cool, but well beyond the lowest common denominator. I think i'ld need to play with it to feel if it can be nice. Even if it can't  it doesn't actually block that first step

Chris Rackauckas [9:51 PM]
or at least lay out a plan that would work for them :slightly_smiling_face:
IMO this is one of the best ways to solve the row/column problem
if things know what kind of data they are reading in, then they can transpose as they please.
But you know how your data is going to be structured in advance, so in that sense you know that if you ask for it differently then someone has to transform it.
either the user or internally.

Dave F Kleinschmidt [9:53 PM]
I wonder if there will be performance gotchas if a modeling package thinks its getting data with a "fast" memory layout but it's really got a view

Chris Rackauckas [9:53 PM]
`collect` is a no-op
on an array

Lyndon White [9:54 PM]
Is it? I thought it was a copy?

Dave F Kleinschmidt [9:54 PM]
even for a `ReshapedArray`?

Chris Rackauckas [9:54 PM]
is it a copy?
oh okay
make a no-op collect then :stuck_out_tongue:
I wonder why I thought it was.
>I wonder if there will be performance gotchas if a modeling package thinks its getting data with a "fast" memory layout but it's really got a view

With lazy transpose, yes.
it's both a blessing and a curse
lazy transpose speeds up a lot of things, and slows down others.

Dave F Kleinschmidt [9:56 PM]
yeah, but it's also encoded by type info so it's mangaeable in package code if necessary

Chris Rackauckas [9:56 PM]
but the fact that it's a choice now and works with the dispatch system is awesome

Dave F Kleinschmidt [9:56 PM]
or you can generate warnings

Chris Rackauckas [9:57 PM]
yes, it has type info which is key. (edited)

Lyndon White [9:57 PM]
And way, it is late. I'll review thus tomorrow, and make an issue. With notes.
And then I'll implement it in fork  on as many packages as I can (edited)
Maybe make a demo
Then try and sell it to all the maintainers and make the PRs

Lyndon White [12:11 PM]
There is a much less minimal API I can imagine working out.
It has some attractions to but also seems over engineered (edited)
Use traits to say which dimensions a classifier uses as obversations
and Use traits to say if it is a BinaryClassifier, or an a MultiClassClassifier etc
then you can do a bunch of nice stuff
This seems close to just having LearnBase.jl as the core package.
So maybe I'll try that if minimal ModelingBase doesn't work out
